### DISCUSSION TITLE
Gpu memory increase

### DISCUSSION THREAD

**Discussion by assistechsol** [mindee/doctr#917] [Q&A]
i am using tensorflow model of DocTr and during prediction GPU memory increase incrementally

**Comment by felixdittrich92** [comment]
Hi @assistechsol,
Could you please share a bit more information about your environment and maybe a code snippet you have used ? :)
cc @charlesmindee

**Reply by assistechsol** [reply]
i am using tensorflow

```
os.environ['USE_TF']= '1'
import tensorflow as tf
from doctr.models import ocr_predictor
os.environ["CUDA_VISIBLE_DEVICES"]="1"
gpu_devices = tf.config.experimental.list_physical_devices('GPU')
if len(gpu_devices) > 0:
         tf.config.experimental.set_visible_devices(gpu_devices[0], 'GPU')
         tf.config.experimental.set_memory_growth(gpu_devices[0], True)
         tf.config.experimental.set_memory_growth(gpu_devices[1], True)

predict = ocr_predictor(pretrained=True)
```

**Reply by frgfm** [reply]
Hi @assistechsol 👋 

Let's try to help you:
- could you share your full snippet where you notice GPU RAM increasing please? This snippet is only about the model initialization
- you used `tf.config.experimental.set_memory_growth` which I wouldn't recommend if you're trying to achieving stable memory usage 😅

**Reply by assistechsol** [reply]
code which i used
```python
predict = ocr_predictor(pretrained=True,
                          det_bs=4,
                          reco_bs=128).cuda()
```
I am using gunicorn workers (5 workers)

**starting gpu memory when model loads**

![Screenshot from 2022-05-17 12-49-25](https://user-images.githubusercontent.com/92520752/168759057-78750f1b-5622-44bf-b3ac-25aca0785764.png)

**During prediction gpu memory increased**

![Screenshot from 2022-05-17 12-50-25](https://user-images.githubusercontent.com/92520752/168759830-d17ce93c-33f7-43b8-9304-f6c3c848212f.png)

but during prediction Gpu memory is not remain same, it is incrementally increase

![Screenshot from 2022-05-17 12-52-26](https://user-images.githubusercontent.com/92520752/168759946-81631231-f33d-4732-9b84-adc189ece259.png)

**Reply by assistechsol** [reply]
I replace tensor flow code to pytorch

**Reply by frgfm** [reply]
Again, the snippet you just shared is only instantiating a model. While the process on your GPU points towards FastAPI. Are you using our API template?

We need to check the code used for inference if we are to help you 😅

**Reply by assistechsol** [reply]
**Here we load model**

![load_model](https://user-images.githubusercontent.com/92520752/168959047-66d6ac2b-43fe-438d-bc91-eb344428a584.png)

**And this portion is used for prediction**

![Screenshot from 2022-05-18 09-44-10](https://user-images.githubusercontent.com/92520752/168959152-c1ae73b5-5aa2-464d-b5dd-3a6d6b70d86a.png)

**Reply by felixdittrich92** [reply]
Hi @assistechsol 
That's only for testing it will decrease the performance but can you please set torch.cuda.empty_cache() after response = predict([image]) and track the mem usage again and share with us ? :)

**Reply by frgfm** [reply]
I'd suggest something extra:
- if you don't necessarily need GPU, running an API on GPU might be tricky
- I don't know how you ran the uvicorn process for fastAPI but it looks like you have 5. If you launched several workers, it's as if you had N models running (where N is the number of workers). That could help
- for a single worker, it looks like it can go up to 2149Mb of GPU RAM. The first picture you shared was when the model were initiated, for orchestration, you need to give models on GPU a few warm-up runs. If after those warmups, the memory is still increasing, then there might be a memory leak 🤔
